File: sdks/python/vectorless/_retry.py
from __future__ import annotations

import asyncio
import random
import time
from typing import Callable, TypeVar, Awaitable

from vectorless.errors import VectorlessError, RateLimitError

T = TypeVar("T")


def _jitter() -> float:
    return random.random() * 0.2


def sync_retry(
    fn: Callable[[], T],
    max_retries: int = 3,
    retry_delay: float = 0.5,
) -> T:
    last_error: BaseException | None = None

    for attempt in range(max_retries + 1):
        try:
            return fn()
        except VectorlessError as e:
            last_error = e

            if e.status == 429:
                delay = (
                    e.retry_after  # type: ignore[union-attr]
                    if isinstance(e, RateLimitError) and e.retry_after is not None
                    else retry_delay * (2**attempt)
                )
                if attempt < max_retries:
                    time.sleep(delay + _jitter())
                    continue

            if 400 <= e.status < 500 and e.status not in (408, 429):
                raise

            if attempt < max_retries:
                time.sleep(retry_delay * (2**attempt) + _jitter())
                continue

            raise
        except Exception as e:
            last_error = e
            if attempt < max_retries:
                time.sleep(retry_delay * (2**attempt) + _jitter())
                continue
            raise

    raise last_error  # type: ignore[misc]


async def async_retry(
    fn: Callable[[], Awaitable[T]],
    max_retries: int = 3,
    retry_delay: float = 0.5,
) -> T:
    last_error: BaseException | None = None

    for attempt in range(max_retries + 1):
        try:
            return await fn()
        except VectorlessError as e:
            last_error = e

            if e.status == 429:
                delay = (
                    e.retry_after  # type: ignore[union-attr]
                    if isinstance(e, RateLimitError) and e.retry_after is not None
                    else retry_delay * (2**attempt)
                )
                if attempt < max_retries:
                    await asyncio.sleep(delay + _jitter())
                    continue

            if 400 <= e.status < 500 and e.status not in (408, 429):
                raise

            if attempt < max_retries:
                await asyncio.sleep(retry_delay * (2**attempt) + _jitter())
                continue

            raise
        except Exception as e:
            last_error = e
            if attempt < max_retries:
                await asyncio.sleep(retry_delay * (2**attempt) + _jitter())
                continue
            raise

    raise last_error  # type: ignore[misc]

File: packages/ts-sdk/src/retry.ts
import type { VectorlessConfig } from "./config.js";
import { VectorlessError, RateLimitError } from "./errors.js";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function jitter(): number {
  return Math.random() * 200;
}

export async function withRetry<T>(
  fn: () => Promise<T>,
  config: Pick<VectorlessConfig, "maxRetries" | "retryDelay">
): Promise<T> {
  let lastError: unknown;

  for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;

      if (error instanceof VectorlessError) {
        // Rate limit: respect Retry-After header
        if (error.status === 429) {
          const delay =
            error instanceof RateLimitError && error.retryAfter != null
              ? error.retryAfter * 1000
              : config.retryDelay * Math.pow(2, attempt);
          if (attempt < config.maxRetries) {
            await sleep(delay + jitter());
            continue;
          }
        }

        // Don't retry 4xx errors (except 408 timeout and 429 rate limit)
        if (
          error.status >= 400 &&
          error.status < 500 &&
          error.status !== 408 &&
          error.status !== 429
        ) {
          throw error;
        }
      }

      // Retry 5xx, network errors, timeouts
      if (attempt < config.maxRetries) {
        await sleep(config.retryDelay * Math.pow(2, attempt) + jitter());
        continue;
      }

      throw lastError;
    }
  }

  throw lastError;
}
